File: 05-軟體工程/03-品質/js/08-floatbug/diff.js
// n 次微分 : 參考 https://en.wikipedia.org/wiki/Finite_difference
function diffn(f, n, x=null, h=0.01) {
  if (n === 0) return f(x)
  h = h/2 // 讓 1, 2, .... n 次微分的最大距離都一樣
  var x1 = (x==null) ? null : x+h
  var x_1 = (x==null) ? null : x-h
  return (diffn(f,n-1,x1) - diffn(f,n-1,x_1))/(2*h)
}

let sin = Math.sin

for (let i=0; i<15; i++) {
  console.log('diffn(sin, %d, PI/3)=%d', i, diffn(sin, i, Math.PI/3))
}

File: 00-程式與數學/03-微積分/01d-微積分基本定理/diff.py
import math

def diff(f, x, dx = 0.0001) :
    print('f(x)=', f(x))
    print('f(x+dx)=', f(x+dx))
    dy = f(x + dx) - f(x) # dy = f(x + dx) - f(x - dx)
    return dy/dx          # return dy / (dx + dx)

if __name__=="__main__":
    print('diff(sin(x), pi/3) = ', diff(math.sin, math.pi / 3))
    print('cos(pi/3) = ', math.cos(math.pi / 3))
